src/class/basic.rs

```diff
@@ -133,94 +133,109 @@ pub trait PyObjectRichcmpProtocol<'p>: PyObjectProtocol<'p> {
     type Result: IntoPyCallbackOutput<PyObject>;
 }
 
-/// All FFI functions for basic protocols.
-#[derive(Default)]
-pub struct PyObjectMethods {
-    pub tp_str: Option<ffi::reprfunc>,
-    pub tp_repr: Option<ffi::reprfunc>,
-    pub tp_hash: Option<ffi::hashfunc>,
-    pub tp_getattro: Option<ffi::getattrofunc>,
-    pub tp_richcompare: Option<ffi::richcmpfunc>,
-    pub tp_setattro: Option<ffi::setattrofunc>,
-    pub nb_bool: Option<ffi::inquiry>,
-}
-
+/// Extension trait for proc-macro backend.
 #[doc(hidden)]
-impl PyObjectMethods {
-    // Set functions used by `#[pyproto]`.
-    pub fn set_str<T>(&mut self)
+pub trait PyBasicSlots {
+    fn get_str() -> ffi::PyType_Slot
```

Comment:
If you take my suggestion above then this should just return the typed function pointer:

```suggestion
    fn get_str() -> ffi::reprfunc {
        py_unary_func!(PyObjectStrProtocol, Self::__str__)
    }
```

And same for all other methods. I think this'll reduce the LOC quite a lot!